central/serviceaccount/datastore/datastore_sac_test.go

```diff
@@ -0,0 +1,375 @@
+		"full read-write on matching cluster cannot remove": {
+			scopeKey:    testutils.Cluster2ReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"read-write on the matching cluster and namespace cannot remove": {
+			scopeKey:    testutils.Cluster2NamespaceBReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+		"read-write on the matching cluster and at least the right namespace cannot remove": {
+			scopeKey:    testutils.Cluster2NamespacesABReadWriteCtx,
+			expectFail:  true,
+			expectedErr: sac.ErrResourceAccessDenied,
+		},
+	}
+
+	for name, c := range cases {
+		s.Run(name, func() {
+			role := fixtures.GetScopedServiceAccount(uuid.NewV4().String(), testconsts.Cluster2,
+				testconsts.NamespaceB)
+			s.testServiceAccountIDs = append(s.testServiceAccountIDs, role.GetId())
+
+			ctx := s.testContexts[c.scopeKey]
+			err := s.datastore.UpsertServiceAccount(s.testContexts[testutils.UnrestrictedReadWriteCtx], role)
+			s.Require().NoError(err)
+			defer s.deleteServiceAccount(role.GetId())
+
+			err = s.datastore.RemoveServiceAccount(ctx, role.GetId())
+			if c.expectFail {
+				s.Require().Error(err)
+				s.ErrorIs(err, c.expectedErr)
+			} else {
+				s.NoError(err)
+			}
+		})
+	}
+}
+
+func (s *serviceAccountSACSuite) runSearchRawTest(c testutils.SACSearchTestCase) {
+	ctx := s.testContexts[c.ScopeKey]
+	results, err := s.datastore.SearchRawServiceAccounts(ctx, nil)
+	s.Require().NoError(err)
+	resultObjs := make([]sac.NamespaceScopedObject, 0, len(results))
+	for i := range results {
+		resultObjs = append(resultObjs, results[i])
+	}
+	resultCounts := testutils.CountSearchResultObjectsPerClusterAndNamespace(s.T(), resultObjs)
+	testutils.ValidateSACSearchResultDistribution(&s.Suite, c.Results, resultCounts)
+}
+
+func (s *serviceAccountSACSuite) runSearchTest(c testutils.SACSearchTestCase) {
+	ctx := s.testContexts[c.ScopeKey]
+	results, err := s.datastore.Search(ctx, nil)
+	s.Require().NoError(err)
+	resultCounts := testutils.CountResultsPerClusterAndNamespace(s.T(), results, s.optionsMap)
+	testutils.ValidateSACSearchResultDistribution(&s.Suite, c.Results, resultCounts)
+}
+
+func (s *serviceAccountSACSuite) TestScopedSearch() {
+	for name, c := range testutils.GenericScopedSACSearchTestCases(s.T()) {
+		s.Run(name, func() {
+			s.runSearchTest(c)
+		})
+	}
+}
+
+func (s *serviceAccountSACSuite) TestUnrestrictedSearch() {
+	for name, c := range testutils.GenericUnrestrictedSACSearchTestCases(s.T()) {
+		s.Run(name, func() {
+			s.runSearchTest(c)
+		})
+	}
+}
+
+func (s *serviceAccountSACSuite) TestScopeSearchRaw() {
+	for name, c := range testutils.GenericScopedSACSearchTestCases(s.T()) {
+		s.Run(name, func() {
+			s.runSearchRawTest(c)
+		})
+	}
+}
+
+func (s *serviceAccountSACSuite) TestUnrestrictedSearchRaw() {
+	for name, c := range testutils.GenericUnrestrictedRawSACSearchTestCases(s.T()) {
+		s.Run(name, func() {
+			s.runSearchRawTest(c)
+		})
+	}
+}
```

Comment:
I've added tests for both.

One concern I had within the `Count` test:
I expected that with different scopes i.e. `Cluster1ReadWriteCtx` the `Count` would yield 9. Instead, this will lead to a panic since we don't have full access to the resource and try to search first and get the length of results.
However, [ApplyCount](https://github.com/stackrox/stackrox/blob/7d33aed81bfbdd59028193427258d7be3f866a5b/pkg/sac/search_helper.go#L85-L93) will set the search func to nil, which will lead to a [nil-pointer panic within this function](https://github.com/stackrox/stackrox/blob/7d33aed81bfbdd59028193427258d7be3f866a5b/pkg/sac/search_helper.go#L147). I haven't figured out how to properly test this within the integration tests, I'll take a look on Monday.

In the mean time, do you have any idea how to do this properly? Haven't found _any_ test that's using count (except for the ones within the `store/postgres`, which are fundamentally different.

Reply:
Looking for code calls:
`ApplyCount` is used in `processindicator`, `rbac/k8srole`, `rbac/k8srolebinding` and `serviceaccount`. Either the associated datastores do not expose any `Count` method, or there are no direct calls to the exposed datastore `Count` functions. I'm not sure how the graphql layer binds the datastores and which functions are used there.

Reply:
Hm, weirdly enough, the `Count` method of the service account's datastore doesn't seem to be used anywhere, which is weird.

I suppose I leave the tests for now and use the unrestricted scope for counting.